src/plugins/expressions/common/service/expressions_services.ts

```diff
@@ -0,0 +1,160 @@
+   * }
+   * ```
+   *
+   * The actual function is defined in the `fn` key. The function can be *async*.
+   * It receives three arguments: (1) `input` is the output of the previous function
+   * or the initial input of the expression if the function is first in chain;
+   * (2) `args` are function arguments as defined in expression string, that can
+   * be edited by user (e.g in case of Canvas); (3) `context` is a shared object
+   * passed to all functions that can be used for side-effects.
+   */
+  public readonly registerFunction = (
+    ...args: Parameters<Executor['registerFunction']>
+  ): ReturnType<Executor['registerFunction']> => this.executor.registerFunction(...args);
+
+  /**
+   * Executes expression string or a parsed expression AST and immediately
+   * returns the result.
+   *
+   * Below example will execute `sleep 100 | clog` expression with `123` initial
+   * input to the first function.
+   *
+   * ```ts
+   * expressions.run('sleep 100 | clog', 123);
+   * ```
+   *
+   * - `sleep 100` will delay execution by 100 milliseconds and pass the `123` input as
+   *   its output.
+   * - `clog` will print to console `123` and pass it as its output.
+   * - The final result of the execution will be `123`.
+   *
+   * Optionally, you can pass an object as the third argument which will be used
+   * to extend the `ExecutionContext`&mdash;an object passed to each function
+   * as the third argument, that allows functions to perform side-effects.
+   *
+   * ```ts
+   * expressions.run('...', null, { elasticsearchClient });
+   * ```
+   */
+  public readonly run = <
+    Input,
+    Output,
+    ExtraContext extends Record<string, unknown> = Record<string, unknown>
+  >(
+    ast: string | ExpressionAstExpression,
+    input: Input,
+    context?: ExtraContext
+  ): Promise<Output> => this.executor.run<Input, Output, ExtraContext>(ast, input, context);
+
+  public setup() {
+    const { executor, renderers, registerFunction, run } = this;
+
+    const getFunctions = executor.getFunctions.bind(executor);
+    const getRenderer = renderers.get.bind(renderers);
+    const getRenderers = renderers.toJS.bind(renderers);
+    const getTypes = executor.getTypes.bind(executor);
+    const registerRenderer = renderers.register.bind(renderers);
+    const registerType = executor.registerType.bind(executor);
+
+    return {
+      getFunctions,
+      getRenderer,
+      getRenderers,
+      getTypes,
+      registerFunction,
+      registerRenderer,
+      registerType,
+      run,
+    };
+  }
+
+  public start() {
+    const { executor, renderers, run } = this;
+
+    const getFunctions = executor.getFunctions.bind(executor);
+    const getRenderer = renderers.get.bind(renderers);
+    const getRenderers = renderers.toJS.bind(renderers);
+    const getTypes = executor.getTypes.bind(executor);
+
+    return {
+      getFunctions,
+      getRenderer,
+      getRenderers,
+      getTypes,
+      run,
+    };
+  }
+
+  public stop() {}
+}
```

Comment:
Awesome that you were able to share the service definition between client & server ⚡️